typescript/core/src/parse.ts

```diff
@@ -1,374 +1,417 @@
+      break;
+    case Opcode.ATTACHMENT_INDEX:
+      result = parseAttachmentIndex(reader, recordLengthNum);
+      break;
+    case Opcode.STATISTICS:
+      result = parseStatistics(reader, recordLengthNum);
+      break;
+    case Opcode.METADATA:
+      result = parseMetadata(reader, recordLengthNum);
+      break;
+    case Opcode.METADATA_INDEX:
+      result = parseMetadataIndex(reader, recordLengthNum);
+      break;
+    case Opcode.SUMMARY_OFFSET:
+      result = parseSummaryOffset(reader, recordLengthNum);
+      break;
+    case Opcode.DATA_END:
+      result = parseDataEnd(reader, recordLengthNum);
+      break;
+    default:
+      result = parseUnknown(reader, recordLengthNum, opcode);
+      break;
   }
 
-  const recordView = new DataView(
-    view.buffer,
-    view.byteOffset + headerReader.offset,
-    recordLengthNum,
+  // NOTE: a bit redundant, but ensures we've advanced by the full record length
+  // TODO: simplify this when we explore monomorphic paths
+  reader.offset = start + RECORD_HEADER_SIZE + recordLengthNum;
+
+  return result;
+}
+
+function parseUnknown(reader: Reader, recordLength: number, opcode: number): TypedMcapRecord {
+  const data = reader.u8ArrayBorrow(recordLength);
+  return {
+    type: "Unknown",
+    opcode,
+    data,
+  };
+}
+
+function parseHeader(reader: Reader, recordLength: number): TypedMcapRecord {
+  const startOffset = reader.offset;
+  const profile = reader.string();
+  const library = reader.string();
+  reader.offset = startOffset + recordLength;
+  return { type: "Header", profile, library };
+}
+
+function parseFooter(reader: Reader, recordLength: number): TypedMcapRecord {
+  const startOffset = reader.offset;
+  const summaryStart = reader.uint64();
+  const summaryOffsetStart = reader.uint64();
+  const summaryCrc = reader.uint32();
+  reader.offset = startOffset + recordLength;
+  return {
+    type: "Footer",
+    summaryStart,
+    summaryOffsetStart,
+    summaryCrc,
+  };
+}
+
+function parseSchema(reader: Reader, recordLength: number): TypedMcapRecord {
+  const start = reader.offset;
+  const id = reader.uint16();
+  const name = reader.string();
+  const encoding = reader.string();
+  const dataLen = reader.uint32();
+  const end = reader.offset;
+  if (recordLength - (end - start) < dataLen) {
+    throw new Error(`Schema data length ${dataLen} exceeds bounds of record`);
+  }
+  const data = reader.u8ArrayCopy(dataLen);
+  reader.offset = start + recordLength;
+
+  return {
+    type: "Schema",
+    id,
+    encoding,
+    name,
+    data,
+  };
+}
+
+function parseChannel(reader: Reader, recordLength: number): TypedMcapRecord {
+  const startOffset = reader.offset;
+  const channelId = reader.uint16();
+  const schemaId = reader.uint16();
+  const topicName = reader.string();
+  const messageEncoding = reader.string();
+  const metadata = reader.map(
+    (r) => r.string(),
+    (r) => r.string(),
```

Comment:
Would it performance-wise be better to avoid the closures? E.g. something like that:

```ts
function readString(reader: Reader) {
  return reader.string();
}
...

const metadata = reader.map(readString, readString)
```

Reply:
Yes, it'd likely be a small net gain, but it's worth optimizing only/mainly if metadata are in the hot path (like messages).